### PR TITLE
Transactions resubmitter

### LINKED ISSUE #1041: Transaction re-submitter helper
Spin-off of #73 

Currently the relayer does not add any tip at all. This works fine if network block utilisation is low, but may cause issues (front-running) in case of congestion.

Instead of supporting `tip` directly in the relayer, I'd rather propose to write a tool that is able to monitor and periodically re-submit the same transaction with exponentialy-increasing `tip` values (with some reasonable cap and alterting).

Basically the tool should:
1. Be in control of a bunch of private keys
2. Monitor transaction pool (and/or have some way to receive "transaction requests")
3. When the transaction is not going through, or is below the configured threshold (compared to other transactions in the pool) it should take the transaction and re-sign it with higher `tip` value.
4. The strategy about `tip`-increase should be configurable (per sender / per transaction kind?), but initially even linear increase seems good enough.

The tool can also server as a "tip oracle" for the relayer.

### PR DESCRIPTION
**Pull request by svyatonik**
closes #1041 

On real-value chains, the strategy will need to be changed (e.g. instead of bumping priority to make tx best, we may just increased tip by fixed value each block it is stalled), but most of code will stay the same. The main point of this PR now is: (1) to fix problems we have with Millau -> Rialto messages (transactions-with-big-messages are too big and are not 'mined' y Millau if there are any other transactions) and (2) to ensure that the resubmit code works.

### CODE REVIEW

File: relays/bin-substrate/src/cli/resubmit_transactions.rs [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+	target: TargetConnectionParams,
+	#[structopt(flatten)]
+	target_sign: TargetSigningParams,
+}
+
+/// Chain, which transactions we're going to track && resubmit.
+#[derive(Debug, EnumString, EnumVariantNames)]
+#[strum(serialize_all = "kebab_case")]
+pub enum RelayChain {
+	Millau,
```
Comment (tomusdrw) [inline relays/bin-substrate/src/cli/resubmit_transactions.rs:47]: Can we add `Rialto` as well if it's not a big effort?

File: relays/bin-substrate/src/cli/resubmit_transactions.rs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+		}
+
+		return Ok(Some(pending_transaction));
+	}
+
+	Ok(None)
+}
+
+/// Read priority of best signed transaction of previous block.
+async fn read_previous_best_priority<C: Chain, S: TransactionSignScheme<Chain = C>>(
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply relays/bin-substrate/src/cli/resubmit_transactions.rs:246]: Or maybe let's pick a median?

**Review by tomusdrw — APPROVED** [review]

File: relays/bin-substrate/src/cli/resubmit_transactions.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (svyatonik) [reply relays/bin-substrate/src/cli/resubmit_transactions.rs:246]: As I've stated in PR description, this strategy won't work well for chains that have real value tokens - you're right. This strategy introduced here is specifically to solve problem that we have with large Millau -> Rialto messages. The problem is that if there's any other transaction in the pool and we're trying to send maximal-size message from Millau -> Rialto, then this send-message transaction won't fit into the Millau block. So its priority must be larger than any other transaction priority, or it'll be staying in the pool and keep generating alerts until a moment when it'll became the only transaction from the pool. Then it'll be mined and alerts will disappear.

When we'll be sure that this code works and solves our issue, then we may introduce other strategy for other bridges.

File: relays/bin-substrate/src/cli/resubmit_transactions.rs (hunk 1 above)
Comment (svyatonik) [reply relays/bin-substrate/src/cli/resubmit_transactions.rs:47]: I can, but existing strategy (see my other comment) is only suitable for specific chain (Millau) and specific problem which we have. So let's probably wait until it works && we have strategy for real-value chains. Then we may enable it for Rialto and start Rialto resubmitter at our deployments

File: relays/client-millau/src/lib.rs [hunk 3]
```diff
@@ -96,6 +95,30 @@ impl TransactionSignScheme for Millau {
...
+			.as_ref()
+			.map(|(address, _, _)| *address == millau_runtime::Address::from(*signer.public().as_array_ref()))
+			.unwrap_or(false)
+	}
+
+	fn parse_transaction(tx: Self::SignedTransaction) -> Option<UnsignedTransaction<Self::Chain>> {
+		let extra = &tx.signature.as_ref()?.2;
+		Some(UnsignedTransaction {
+			call: tx.function,
+			nonce: Compact::<<Self::Chain as Chain>::Index>::decode(&mut &extra.4.encode()[..])
```
[the remark this answers is not in the document]
Comment (svyatonik) [reply relays/client-millau/src/lib.rs:114]: Unfortunately, signed extensions are not providing any means to access their data. Other option is to `parse(format!("{}", extra.4))`, but it is even worse :)

**Pull request merged**